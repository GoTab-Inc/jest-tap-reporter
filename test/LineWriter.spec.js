/* eslint-disable max-nested-callbacks */
const chalk = require('chalk');
const Logger = require('../src/Logger');
const LineWriter = require('../src/LineWriter');

jest.mock('chalk');
jest.mock('../src/Logger');

const create = (root = '/jest-tap-reporter', logger = new Logger()) => {
  const writer = new LineWriter(logger, root);

  return writer;
};

const MDASH = '\u2014';
const CIRCLE = '●';

describe('LiveWriter', () => {
  test('is a function', () => {
    expect(typeof LineWriter).toBe('function');
  });

  test('can construct without error', () => {
    create();
  });

  describe('.getNextNumber()', () => {
    test('returns a positive number', () => {
      const writer = create();
      const num = writer.getNextNumber();

      expect(typeof num).toBe('number');
      expect(num).toBeGreaterThan(0);
    });

    test('starts with 1', () => {
      const writer = create();
      const num = writer.getNextNumber();

      expect(num).toBe(1);
    });

    test('each next number increases by 1', () => {
      const writer = create();
      let last = writer.getNextNumber();

      for (let j = 0; j < 100; j++) {
        const current = writer.getNextNumber();

        expect(last + 1).toBe(current);
        last = current;
      }
    });
  });

  describe('.blank()', () => {
    test('logs a blank line using .info()', () => {
      const writer = create();

      writer.blank();

      expect(writer.logger.info).toHaveBeenCalledTimes(1);
      expect(writer.logger.info.mock.calls[0]).toEqual(['']);
    });
  });

  describe('.comment()', () => {
    test('logs using .info()', () => {
      const writer = create();

      writer.comment('something');

      expect(writer.logger.info).toHaveBeenCalledTimes(1);
    });

    test('writes a TAP comment', () => {
      chalk.__stripColors();

      const writer = create();

      writer.comment('foo');

      expect(writer.logger.info).toHaveBeenCalledTimes(1);
      expect(writer.logger.info.mock.calls[0]).toEqual(['# foo']);
    });

    test('hides comment hash symbol from console', () => {
      chalk.__showTemplates();

      const writer = create();

      writer.comment('foo');

      expect(writer.logger.info).toHaveBeenCalledTimes(1);
      expect(writer.logger.info.mock.calls[0]).toEqual(['{hidden #} foo']);
    });
  });

  describe('.start()', () => {
    test('prints start message', () => {
      chalk.__stripColors();

      const writer = create();

      writer.start();

      expect(writer.logger.info.mock.calls).toMatchSnapshot();
    });
  });

  describe('.commentLight()', () => {
    test('prints dimmed comment', () => {
      const writer = create();

      writer.comment = jest.fn();
      writer.commentLight('foo');

      expect(writer.comment.mock.calls).toMatchSnapshot();
    });
  });

  describe('.keyValue()', () => {
    test('prints key-value pair', () => {
      const writer = create();

      writer.comment = jest.fn();
      writer.keyValue('foo', 'bar');

      expect(writer.comment.mock.calls).toMatchSnapshot();
    });
  });

  describe('.stats()', () => {
    describe('when zero tests', () => {
      test('shows only total zero', () => {
        const writer = create();

        writer.keyValue = jest.fn();
        writer.stats('foo', 0, 0, 0, 0);

        expect(writer.keyValue).toHaveBeenCalledTimes(1);
        expect(writer.keyValue.mock.calls[0]).toMatchSnapshot();
      });
    });

    describe('when all tests pass', () => {
      test('shows only passed and total tests', () => {
        const writer = create();

        writer.keyValue = jest.fn();
        writer.stats('foo', 0, 0, 1, 1);

        expect(writer.keyValue).toHaveBeenCalledTimes(1);
        expect(writer.keyValue.mock.calls[0]).toMatchSnapshot();
      });
    });

    describe('when some tests fail', () => {
      test('shows only passed, failed and total tests', () => {
        const writer = create();

        writer.keyValue = jest.fn();
        writer.stats('foo', 1, 0, 1, 2);

        expect(writer.keyValue).toHaveBeenCalledTimes(1);
        expect(writer.keyValue.mock.calls[0]).toMatchSnapshot();
      });
    });

    describe('when some tests are skipped', () => {
      test('shows all items', () => {
        const writer = create();

        writer.keyValue = jest.fn();
        writer.stats('foo', 1, 1, 1, 3);

        expect(writer.keyValue).toHaveBeenCalledTimes(1);
        expect(writer.keyValue.mock.calls[0]).toMatchSnapshot();
      });

      describe('and no tests fail', () => {
        test('shows only passed, skipped and total tests', () => {
          const writer = create();

          writer.keyValue = jest.fn();
          writer.stats('foo', 0, 1, 1, 2);

          expect(writer.keyValue).toHaveBeenCalledTimes(1);
          expect(writer.keyValue.mock.calls[0]).toMatchSnapshot();
        });
      });
    });
  });

  describe('.result()', () => {
    test('logs passed test', () => {
      const writer = create();

      writer.logger.log = jest.fn();
      writer.result('ok', 'Test passed');

      expect(writer.logger.log).toHaveBeenCalledTimes(1);
      expect(writer.logger.log.mock.calls[0][0]).toBe('ok 1 Test passed');
    });

    test('logs failed test', () => {
      const writer = create();

      writer.logger.log = jest.fn();
      writer.result('not ok', 'Test failed');

      expect(writer.logger.log).toHaveBeenCalledTimes(1);
      expect(writer.logger.log.mock.calls[0][0]).toBe('not ok 1 Test failed');
    });

    test('increments test counter', () => {
      const writer = create();

      writer.logger.log = jest.fn();

      writer.result('ok', 'Test passed');
      writer.result('not ok', 'Test failed');

      expect(writer.logger.log).toHaveBeenCalledTimes(2);

      expect(writer.logger.log.mock.calls[0][0]).toBe('ok 1 Test passed');
      expect(writer.logger.log.mock.calls[1][0]).toBe('not ok 2 Test failed');
    });
  });

  describe('.passed()', () => {
    test('calls .result() with the right parameters', () => {
      chalk.__stripColors();

      const writer = create();

      writer.result = jest.fn();
      writer.passed('Test passed');

      expect(writer.result).toHaveBeenCalledTimes(1);
      expect(writer.result.mock.calls[0]).toEqual(['ok', `${MDASH} Test passed`]);
    });

    test('colors "ok" green', () => {
      chalk.__showTemplates();

      const writer = create();

      writer.result = jest.fn();
      writer.passed('Test passed');

      expect(writer.result).toHaveBeenCalledTimes(1);
      expect(writer.result.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('.failed()', () => {
    test('calls .result() with the right parameters', () => {
      chalk.__stripColors();

      const writer = create();

      writer.result = jest.fn();
      writer.failed('Test failed');

      expect(writer.result).toHaveBeenCalledTimes(1);
      expect(writer.result.mock.calls[0]).toEqual(['not ok', `${CIRCLE} Test failed`]);
    });

    test('colors "not ok" red', () => {
      chalk.__showTemplates();

      const writer = create();

      writer.result = jest.fn();
      writer.failed('Test failed');

      expect(writer.result).toHaveBeenCalledTimes(1);
      expect(writer.result.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('.pending()', () => {
    test('calls .result() with the right parameters', () => {
      chalk.__stripColors();

      const writer = create();

      writer.result = jest.fn();
      writer.pending('Test pending');

      expect(writer.result).toHaveBeenCalledTimes(1);
      expect(writer.result.mock.calls[0]).toEqual(['ok', '# SKIP Test pending']);
    });

    test('colors "ok" yellow', () => {
      chalk.__showTemplates();

      const writer = create();

      writer.result = jest.fn();
      writer.pending('Test pending');

      expect(writer.result).toHaveBeenCalledTimes(1);
      expect(writer.result.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('.getPathRelativeToRoot()', () => {
    test('returns path relative to root folder', () => {
      const writer = create('/foo');
      const rel = writer.getPathRelativeToRoot('/foo/bar');

      expect(rel).toBe('bar');
    });
  });

  describe('.suite()', () => {
    test('prints test suite result', () => {
      chalk.__stripColors();

      const writer = create('/foo');

      writer.comment = jest.fn();
      writer.suite(true, '/foo/bar/', 'kappa.js');

      expect(writer.comment).toHaveBeenCalledTimes(1);
      expect(writer.comment.mock.calls[0][0]).toMatchSnapshot();
    });

    test('makes directory path relative', () => {
      chalk.__stripColors();

      const writer = create('/foo');

      writer.comment = jest.fn();
      writer.suite(true, '/foo/bar/', 'kappa.js');

      expect(writer.comment).toHaveBeenCalledTimes(1);

      const line = writer.comment.mock.calls[0][0];

      expect(line.includes('/foo/bar/kappa.js')).toBe(false);
      expect(line.includes('bar/kappa.js')).toBe(true);
    });
  });

  describe('plan', () => {
    test('prints test plan with supplied test count', () => {
      chalk.__stripColors();

      const writer = create();

      writer.logger.log = jest.fn();
      writer.plan(100);

      expect(writer.logger.log).toHaveBeenCalledTimes(1);
      expect(writer.logger.log.mock.calls[0][0]).toBe('1..100');
    });

    test('can be written only once', () => {
      const writer = create();

      writer.logger.log = jest.fn();
      writer.plan(100);
      expect(() => writer.plan(100)).toThrow(Error);
    });

    describe('when test count is not provided', () => {
      test('prints test plan using the current counter as test count', () => {
        chalk.__stripColors();

        const writer = create();

        writer.logger.log = jest.fn();
        writer.result('ok');
        writer.result('ok');
        writer.result('ok');
        writer.plan();

        expect(writer.logger.log).toHaveBeenCalledTimes(4);
        expect(writer.logger.log.mock.calls[3][0]).toBe('1..3');
      });
    });
  });
});
