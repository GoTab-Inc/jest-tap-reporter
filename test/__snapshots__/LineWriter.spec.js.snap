// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveWriter .commentLight() prints dimmed comment 1`] = `
Array [
  Array [
    "foo",
  ],
]
`;

exports[`LiveWriter .errors() format stack trace 1`] = `
"#   
#   foobar
#   
#   Stack trace:
#   
#       at Something (../foo/bar.js:10:10)
#       at Foobar (../foo/bar2.js:20:20)
#   "
`;

exports[`LiveWriter .failed() colors "not ok" red 1`] = `"{red not ok}"`;

exports[`LiveWriter .keyValue() prints key-value pair 1`] = `
Array [
  Array [
    "foo:         bar",
  ],
]
`;

exports[`LiveWriter .passed() colors "ok" green 1`] = `"{green ok}"`;

exports[`LiveWriter .pending() colors "ok" yellow 1`] = `"{yellow ok}"`;

exports[`LiveWriter .snapshots() when all values are greater than zero prints them all 1`] = `
Array [
  "Snapshots",
  "1 failed, 1 updated, 1 added, 1 passed, 4 total",
]
`;

exports[`LiveWriter .start() prints start message 1`] = `
Array [
  Array [
    "",
  ],
  Array [
    "",
  ],
  Array [
    "# Starting...",
  ],
]
`;

exports[`LiveWriter .stats() when all tests pass shows only passed and total tests 1`] = `
Array [
  "foo",
  "1 passed, 1 total",
]
`;

exports[`LiveWriter .stats() when some tests are skipped and no tests fail shows only passed, skipped and total tests 1`] = `
Array [
  "foo",
  "1 skipped, 1 passed, 2 total",
]
`;

exports[`LiveWriter .stats() when some tests are skipped shows all items 1`] = `
Array [
  "foo",
  "1 failed, 1 skipped, 1 passed, 3 total",
]
`;

exports[`LiveWriter .stats() when some tests fail shows only passed, failed and total tests 1`] = `
Array [
  "foo",
  "1 failed, 1 passed, 2 total",
]
`;

exports[`LiveWriter .stats() when zero tests shows only total zero 1`] = `
Array [
  "foo",
  "0 total",
]
`;

exports[`LiveWriter .suite() prints test suite result 1`] = `" FAIL  bar/kappa.js"`;
